### DISCUSSION TITLE
Experimenting with GPT-4

### DISCUSSION THREAD

**Discussion by vsg-dev** [vsg-dev/VulkanSceneGraph#762] [General]
Here's a couple of questions I've tried so far:

**RO: What is a scene graph?**

A scene graph is a data structure used in computer graphics and 3D modeling to represent a hierarchical organization of objects in a scene.

A scene graph typically contains nodes that represent various elements of the scene, such as geometry, lights, cameras, and transformations. These nodes are connected together in a tree-like structure, where each node may have one or more child nodes.

The structure of the scene graph reflects the relationships between the objects in the scene. For example, a car model might have a parent node representing the car body, with child nodes representing the wheels, seats, and steering wheel.

Scene graphs are used in a variety of applications, including video games, virtual reality, and animation. They provide a convenient way to manipulate the objects in a scene and to apply transformations such as translation, rotation, and scaling. Additionally, scene graphs can be used to optimize rendering performance by culling objects that are not visible to the camera or by grouping objects with similar rendering properties.

**What are the pros and cons of the VulkanSceneGraph?**

The VulkanSceneGraph (VSG) is a scene graph library for the Vulkan graphics API. Like any software library, VSG has its own set of pros and cons, which I will outline below:

Pros:

    High performance: VSG is designed to be highly efficient and scalable, allowing it to handle large, complex scenes with ease. This is achieved by leveraging the power of Vulkan's low-level API and minimizing CPU overhead.

    Cross-platform: VSG is designed to be cross-platform, meaning it can run on multiple operating systems and hardware architectures. This makes it a versatile choice for developers who want to target a wide range of platforms.

    Flexibility: VSG provides a flexible framework that allows developers to customize and extend it to fit their specific needs. This includes support for custom shaders, materials, and other graphics components.

    Open source: VSG is an open-source project, which means that it is freely available for anyone to use and contribute to. This has led to a vibrant community of developers who are actively improving and expanding the library.

Cons:

    Steep learning curve: VSG is a complex library that requires a deep understanding of both Vulkan and graphics programming in general. This can make it difficult for beginners to get started.

    Lack of documentation: While VSG is actively maintained, its documentation is not as comprehensive as other graphics libraries. This can make it challenging to understand the library's features and how to use them effectively.

    Limited tooling: Because VSG is a relatively new library, there are not as many third-party tools and plugins available for it as there are for more established libraries. This can make it more challenging to integrate VSG into an existing development pipeline.

    Vulkan-only: As a library designed specifically for Vulkan, VSG is not compatible with other graphics APIs such as OpenGL or DirectX. This means that developers who want to use VSG will need to be familiar with Vulkan and its programming model.

**RO: Which is pretty impressive. I'm working on the tutorial book right now, so I asked it  :  What is the vsgTutorial? **

vsgTutorial is a collection of code examples and tutorials designed to help developers learn how to use the VulkanSceneGraph (VSG) library. It covers a wide range of topics, from basic concepts such as creating a window and rendering a triangle, to more advanced topics such as multithreading and scene culling.

The vsgTutorial is available as part of the VSG repository on GitHub and is open source, allowing developers to study and modify the code as needed. It is designed to be accessible to developers with varying levels of experience, and each tutorial includes detailed explanations of the concepts and code used.

The vsgTutorial is organized into a series of chapters, with each chapter covering a specific topic. Each chapter contains multiple examples, with each example building upon the concepts and code introduced in the previous example.

Some of the topics covered in the vsgTutorial include:

    Creating a Vulkan instance and surface
    Creating a window and handling events
    Creating graphics pipelines and rendering triangles
    Creating vertex and index buffers
    Loading and rendering models
    Implementing multithreading for improved performance
    Implementing scene culling for efficient rendering

Overall, the vsgTutorial is a valuable resource for developers who want to learn how to use the VulkanSceneGraph library to create high-performance graphics applications.

**RO: Which is more wishful thinking at this point as I've only finished a draft of the first chapter so far and haven't yet tackled any of the above topics!**

**Comment by vsg-dev** [comment]
I have tried getting Chat-GPT4 to create a VSG hello world example and while the code looked "plausible" and got some parts right it made lots of errors in how to do very basic things, and even looked to be conflating something you might do with the OSG rather than the VSG, but likely also stuff it's crawled off other materials.  

In the I gave up as no matter how much I tried to coax it into do something valid, while it GPT4 might be capable of writing some code examples it was hopeless at doing something helpful for a VSG application.  Perhaps over time it'll have more VSG code out there is crawl and mimic.

I have been more success on getting to to spot my typos and bad grammar in the online vsgTutorial book I've just started writing.  For this I need every help I can get.  To paraphrase Michael Jackson, "I'm coder not a writer."

**Reply by ghost** [reply]
gpt-4o can't help me as well :) can't even tell me how to create HUD, even provided with the vsg source code
it's hard for me to use vsg at this point of my programming journey, so any hint is appreciated, i mean anything, even the simplest stuff, thanks

**Reply by robertosfield** [reply]
I'd start by creating a thread on the specific topic here on the VSG forum - just ask the question, hopefully one of us humans in the community will be able to point you in the right direction.  

This thread to just AI assisted coding so won't do any general support in this thread otherwise we'll end up a tangled mess of topics.

**Comment by timoore** [comment]
The Chat-GPT4 output seems pretty typical of what I've seen, though I haven't tried to get it to cough up anything myself. It's all grammatically correct. It's done a good job of summarizing the Wikipedia entry on scene graphs, with maybe some other content thrown in. I don't find the pros and cons very helpful; they could apply to just about any graphics library. It's missed what I consider to be the real advantages of VSG:
* OSG cultural compatibility (no mention at all of OSG)
* handles large / ECEF coordinates with ease
* database paging

But those are very nuanced criteria, and Chat-GPT4 just can't pick those signals out of all the Vulkan and API noise.

The statements about the VSG tutorial fall in the category of "hallucinations" (they are actually called that!). You can almost see how it's gotten confused by the actual Vulkan tutorial and all the mentions of Vulkan in VSG materials.

**Reply by vsg-dev** [reply]
I have wrapped up my initial test of Chat-GPT4, I've found it most useful as a glorified spell/grammar checker.  The code and documentation that it can generate is often human readable and to the casual reader make look plausible but gets enough things wrong that it could be very misleading.

**Comment by robertosfield** [comment]
Has anyone experimented with Github Copilot?

Video deep-dive:
https://www.youtube.com/watch?v=8_0DJ9FOlOM

**Comment by mp3butcher** [comment]
###  AI is the future L0L

> how do i deep copy a vsg::Node?

Copilot said: To deep copy a `vsg::Node` in VulkanSceneGraph,

To deep copy a vsg::Node in VulkanSceneGraph, you should use the clone method provided by the vsg::Node class (inherited from vsg::Object). This method uses the copy constructor internally, which takes a CopyOp argument to control the copy behavior. blablabla

> I'm calling clone on a VertexIndexDraw and when called it doesnt deep copy neither arrays neither indices

Copilot said: You are correct to expect a deep copy, because the VertexIndexDraw copy constructor and clone method use the CopyOp mechanism for arrays and indices. However, whether you get a shallow or deep copy depends on the CopyOp you provide to clone()!
The default CopyOp (CopyOp()) performs a shallow copy (simply copies the reference).
 If you want a deep copy (new, duplicated arrays and indices), you must use CopyOp::DEEP_COPY when calling clone . blablabla
 
> where's vsg::CopyOp::DEEP_COPY  is defined?

Copilot said: The symbol vsg::CopyOp::DEEP_COPY does not appear in the top 10 code search results for CopyOp in this repository. In the source code for CopyOp (include/vsg/core/Object.h), there is no enum or static constant named DEEP_COPY visible in the results. The CopyOp class is defined as a functor for copying (with template operator overloads), but no DEEP_COPY constant is shown. blablabla

> so you're telling me non sense... Aren't you?

You are right to question the previous answer—I made an incorrect assumption about the existence of vsg::CopyOp::DEEP_COPY. After searching the VulkanSceneGraph source, there is no such constant or enum named DEEP_COPY in the CopyOp class or elsewhere in the repo. blablabla
